### ISSUE TITLE
BERT Training encounters loss NaN with AMP and TransformerEncoderLayer.normalize_before=True

### ISSUE THREAD

**Issue by mingxu1067** [PaddlePaddle/Paddle#33145]
**System information**
-PaddlePaddle: Both 2.1.0 and b8e4ec7de1e04f1a7fcf6c85ce016f9fef37ee8d
-CPU: Intel(R) Xeon(R) CPU E5-2698 v4 @ 2.20GHz
-GPU: V100-SXM2-16GB with CUDA 11.2 and CUDNN 8.1.1
-OS Platform: Ubuntu 18.04
Docker images: Both registry.baidubce.com/paddlepaddle/paddle:2.1.0-gpu-cuda11.2-cudnn8 and  paddlepaddle/paddle:latest-dev-cuda11.2-cudnn8-gcc82
-Paddle Build Comanad: `cmake .. -DWITH_PYTHON=ON 
         -DPY_VERSION=3.8 
         -DWITH_GPU=ON 
         -DWITH_TESTING=ON 
         -DWITH_INFERENCE_API_TEST=OFF 
         -DCMAKE_BUILD_TYPE=Release 
         -DCUDA_ARCH_NAME=Volta 
         -DON_INFER=OFF 
         -DWITH_MKL=OFF 
         -DWITH_DISTRIBUTE=ON`
-Distributed training: 8xV100-SXM2-16GB
-PaddleNLP:  Both release/2.0-rc and c3d1df6e9e9b77df7af74b7f8a819182ee9c3a7f

**To Reproduce**
-Training Script: https://github.com/PaddlePaddle/PaddleNLP/blob/release/2.0-rc/benchmark/bert/run_pretrain.py
-Turn on normalize_before=True in [paddlenlp/transformer/berts/modeling](https://github.com/PaddlePaddle/PaddleNLP/blob/release/2.0-rc/paddlenlp/transformers/bert/modeling.py#L295))
`encoder_layer = nn.TransformerEncoderLayer(
            hidden_size,
            num_attention_heads,
            intermediate_size,
            dropout=hidden_dropout_prob,
            activation=hidden_act,
            attn_dropout=attention_probs_dropout_prob,
            act_dropout=0,
            normalize_before=True)`
-Training Configuration:
> model type: bert-base-uncased
> sequence lenght: 512
> max predictions per seq: 80 
> batch size per GPU: 16
> learning rate: 1e-4 
> weightdecay: 1e-2 
> adam epsilon: 1e-6 
> warmupsteps: 0 
> use_amp: true 
> use_pure_fp16: true

**Describe your current behavior**
BERT-base pretraining with `normalize_before=True` would led to NaN loss after 7000 update steps. Then it keeps NaN or diverse.

**Other info / logs**
> tobal step: 5000, epoch: 0, batch: 1752, loss: 8.273924, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14668 sec, avg_samples: 16.00000, ips: 109.08294 sequences/sec
> tobal step: 6000, epoch: 0, batch: 2752, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14690 sec, avg_samples: 16.00000, ips: 108.92018 sequences/sec
> tobal step: 7000, epoch: 0, batch: 715, loss: nan, avg_reader_cost: 0.00012 sec, avg_batch_cost: 0.10542 sec, avg_samples: 11.45600, ips: 108.66931 sequences/sec
> tobal step: 8000, epoch: 0, batch: 1715, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14718 sec, avg_samples: 16.00000, ips: 108.70821 sequences/sec
> tobal step: 9000, epoch: 0, batch: 2715, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14719 sec, avg_samples: 16.00000, ips: 108.70467 sequences/sec
> tobal step: 10000, epoch: 0, batch: 604, loss: nan, avg_reader_cost: 0.00010 sec, avg_batch_cost: 0.08906 sec, avg_samples: 9.68000, ips: 108.68977 sequences/sec
> tobal step: 11000, epoch: 0, batch: 1604, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14715 sec, avg_samples: 16.00000, ips: 108.73430 sequences/sec
> tobal step: 12000, epoch: 0, batch: 2604, loss: 8.463623, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14709 sec, avg_samples: 16.00000, ips: 108.77762 sequences/sec
> tobal step: 13000, epoch: 0, batch: 548, loss: 9.715588, avg_reader_cost: 0.00010 sec, avg_batch_cost: 0.08063 sec, avg_samples: 8.78400, ips: 108.94407 sequences/sec
> tobal step: 14000, epoch: 0, batch: 1548, loss: 10.569220, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14693 sec, avg_samples: 16.00000, ips: 108.89843 sequences/sec
> tobal step: 15000, epoch: 0, batch: 2548, loss: 10.297121, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14715 sec, avg_samples: 16.00000, ips: 108.73025 sequences/sec
> tobal step: 16000, epoch: 0, batch: 425, loss: nan, avg_reader_cost: 0.00010 sec, avg_batch_cost: 0.06278 sec, avg_samples: 6.81600, ips: 108.56290 sequences/sec
> tobal step: 17000, epoch: 0, batch: 1425, loss: 10.472240, avg_reader_cost: 0.00006 sec, avg_batch_cost: 0.14708 sec, avg_samples: 16.00000, ips: 108.78791 sequences/sec
> tobal step: 18000, epoch: 0, batch: 2425, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14707 sec, avg_samples: 16.00000, ips: 108.79291 sequences/sec
> tobal step: 19000, epoch: 0, batch: 364, loss: nan, avg_reader_cost: 0.00008 sec, avg_batch_cost: 0.05392 sec, avg_samples: 5.84000, ips: 108.31464 sequences/sec
> tobal step: 20000, epoch: 0, batch: 1364, loss: 10.584019, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14712 sec, avg_samples: 16.00000, ips: 108.75383 sequences/sec
> tobal step: 21000, epoch: 0, batch: 2364, loss: 10.792961, avg_reader_cost: 0.00006 sec, avg_batch_cost: 0.14711 sec, avg_samples: 16.00000, ips: 108.76450 sequences/sec
> tobal step: 22000, epoch: 0, batch: 259, loss: nan, avg_reader_cost: 0.00009 sec, avg_batch_cost: 0.03847 sec, avg_samples: 4.16000, ips: 108.13081 sequences/sec
> tobal step: 23000, epoch: 0, batch: 1259, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14711 sec, avg_samples: 16.00000, ips: 108.76420 sequences/sec
> tobal step: 24000, epoch: 0, batch: 2259, loss: 13.430931, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14657 sec, avg_samples: 16.00000, ips: 109.15941 sequences/sec
> tobal step: 25000, epoch: 0, batch: 69, loss: 15.287128, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.01046 sec, avg_samples: 1.12000, ips: 107.10091 sequences/sec
> tobal step: 26000, epoch: 0, batch: 1069, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14709 sec, avg_samples: 16.00000, ips: 108.77380 sequences/sec
> tobal step: 27000, epoch: 0, batch: 2069, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14710 sec, avg_samples: 16.00000, ips: 108.76978 sequences/sec
> tobal step: 28000, epoch: 0, batch: 3069, loss: nan, avg_reader_cost: 0.00007 sec, avg_batch_cost: 0.14737 sec, avg_samples: 16.00000, ips: 108.57103 sequences/sec
